### DISCUSSION TITLE
Most wanted features

### DISCUSSION THREAD

**Discussion by shaobin** [microsoft/GraphEngine#344] [Ideas]
Please share with us your most wanted Graph Engine features. Thank you!

**Comment by TaviTruman** [comment]
- Support for C# 10/11 Data type. My thinking here is that one be able to specify what type of structures to create during code generation
   - Tuples, C# record struct
   - C# Classes so that DTO can be used with UI (WPF, WinUI, MAUI, etc.) frameworks (XAML)
   - record struct
   - record
   - Generic Types
   - iterate over the in-memory cloud from the client-side
   - Type-safe access to the IStorage (GraphEngineClient) where socket handle is exposed. I've developed a method of doing this so as to enable server-side reactive push.
   - Reactive Event Streaming patterns. I used .NET RX Cold and Hot Observables. This is a  sophisticated topic; not overly complex but does require quite a bit of design pattern work. I've done a great deal of work in this area and have working samples that scale well.
     - Very useful when writing UI and data-driven and event-driven solutions
   - Telemetry API
   - Full ACID support where transactional processing is necessary
   - Graph Engine VSIX for VS 2202 - I am currently porting the existing implementation
   - Visual Studio Templets
      - Protocol-specific support (Server, Proxy, default)
      - Availability (GE Cluster)
      - TSL Graph Data model
      - GE App Server Template
      - Azure Deployment options: (Azure Worker Service, SF, Dapr, etc.,)
   - Graph Engine TSL Visual Editor - I have started work on a VS Add-in that uses the same graphics library (yWorks) that Microsoft uses to generate visual code maps.
   - Adapter Design Patterns to use with LIKQ. ex: indexing, hyper-graph, multi-graph, and graph exploration
   - more built-in Prolog-like primitives (for use with LIKQ)
   - Better integration with Azure 
     - How to deploy GE App Servers and Availability Cluster with Azure VM Scale Set, Service Fabric, Dapr, etc., I currently use Azure Work Service deployed as Windows Service
   - Native F# GE Client, others (Java, Rust)
   - Pattern matching support on GE TSL Attributes

**Reply by TaviTruman** [reply]
I have a lot more but I think is a decent starting point.
